File: src/main/java/com/gmail/borlandlp/minigamesdtools/APIComponentsController.java
package com.gmail.borlandlp.minigamesdtools;

import java.util.ArrayList;
import java.util.List;

public class APIComponentsController {
    private List<APIComponent> components = new ArrayList<>();

    public void register(APIComponent c) {
        this.components.add(c);
    }

    public void unregister(APIComponent c) {
        this.components.remove(c);
    }

    private void announceComponent(APIComponent component, ComponentEvent phase) {
        try {
            switch (phase) {
                case PLUGIN_LOAD: component.onLoad(); break;
                case PLUGIN_UNLOAD: component.onUnload(); break;
                default: throw new Exception("Invalid phase name:" + phase.name());
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public void announceEvent(ComponentEvent componentEvent) {
        for (APIComponent component : this.getComponents()) {
            this.announceComponent(component, componentEvent);
        }
    }

    public List<APIComponent> getComponents() {
        return new ArrayList<>(this.components);
    }

    public enum ComponentEvent {
        PLUGIN_LOAD,
        PLUGIN_UNLOAD,
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/APIComponentsController.kt
package com.gmail.borlandlp.minigamesdtools

import java.util.*

class APIComponentsController {
    private val components: MutableList<APIComponent> = ArrayList()
    fun register(c: APIComponent) {
        components.add(c)
    }

    fun unregister(c: APIComponent) {
        components.remove(c)
    }

    private fun announceComponent(
        component: APIComponent,
        phase: ComponentEvent
    ) {
        try {
            when (phase) {
                ComponentEvent.PLUGIN_LOAD -> component.onLoad()
                ComponentEvent.PLUGIN_UNLOAD -> component.onUnload()
                else -> throw Exception("Invalid phase name:" + phase.name)
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun announceEvent(componentEvent: ComponentEvent) {
        for (component in getComponents()) {
            announceComponent(component, componentEvent)
        }
    }

    fun getComponents(): List<APIComponent> {
        return ArrayList(components)
    }

    enum class ComponentEvent {
        PLUGIN_LOAD, PLUGIN_UNLOAD
    }
}